compiler/noirc_evaluator/src/ssa/ir/dfg.rs

```diff
@@ -486,21 +558,91 @@ impl DataFlowGraph {
         destination.set_terminator(terminator);
     }
 
-    pub(crate) fn get_call_stack(&self, instruction: InstructionId) -> CallStack {
+    pub(crate) fn get_instruction_call_stack(&self, instruction: InstructionId) -> CallStack {
+        let call_stack = self.locations.get(&instruction).cloned().unwrap_or_default();
+        self.get_call_stack(call_stack)
+    }
+
+    pub(crate) fn get_instruction_call_stack_id(&self, instruction: InstructionId) -> CallStackId {
         self.locations.get(&instruction).cloned().unwrap_or_default()
     }
 
-    pub(crate) fn add_location(&mut self, instruction: InstructionId, location: Location) {
-        self.locations.entry(instruction).or_default().push_back(location);
+    pub(crate) fn add_location_to_instruction(
+        &mut self,
+        instruction: InstructionId,
+        location: Location,
+    ) {
+        let call_stack = self.locations.entry(instruction).or_default();
+        call_stack.add_location(location, &mut self.location_array);
+    }
+
+    pub(crate) fn add_location_to_root(&mut self, location: Location) -> CallStackId {
+        if self.location_array.is_empty() {
+            self.location_array.push(LocationNode {
+                parent: None,
+                children: vec![],
+                value: location,
+            });
```

Comment:
I think the fact that here we have to think about the array being empty is a sign that the array itself should be a _newtype_ that holds these methods, like `self.location_array.add_location_to_root` would establish the root ID, and then `self.location_array.add_location(parent_id, location)` would be an example of adding more stuff, rather than the IDs manipulating the collection.

Reply:
I mean the sign that something is not right is that a person cannot do `CallStackId::root().add_location(location, &mut self.location_array)` because it might panic, so `CallStackId` alone is not responsible for the invariants, although it could special case `add_location` for when `self.is_root()` is true and the target is empty.

Reply:
I tried to refactor this.